src/cluster.c

```diff
@@ -4647,31 +4838,56 @@ void clusterUpdateSlots(client *c, unsigned char *slots, int del) {
     }
 }
 
-void addNodeReplyForClusterSlot(client *c, clusterNode *node, int start_slot, int end_slot) {
-    int i, nested_elements = 3; /* slots (2) + master addr (1) */
-    void *nested_replylen = addReplyDeferredLen(c);
-    addReplyLongLong(c, start_slot);
-    addReplyLongLong(c, end_slot);
-    addReplyArrayLen(c, 3);
-    addReplyBulkCString(c, node->ip);
+void addNodeToNodeReply(client *c, clusterNode *node) {
+    addReplyArrayLen(c, 4);
+    if (server.cluster_preferred_endpoint_type == CLUSTER_ENDPOINT_TYPE_IP) {
+        addReplyBulkCString(c, node->ip);
+    } else if (server.cluster_preferred_endpoint_type == CLUSTER_ENDPOINT_TYPE_HOSTNAME) {
+        addReplyBulkCString(c, node->hostname ? node->hostname : "?");
+    } else if (server.cluster_preferred_endpoint_type == CLUSTER_ENDPOINT_TYPE_UNKNOWN_ENDPOINT) {
+        addReplyNull(c);
+    } else {
+        serverPanic("Unrecognized preferred endpoint type");
+    }
+    
     /* Report non-TLS ports to non-TLS client in TLS cluster if available. */
     int use_pport = (server.tls_cluster &&
                      c->conn && connGetType(c->conn) != CONN_TYPE_TLS);
     addReplyLongLong(c, use_pport && node->pport ? node->pport : node->port);
     addReplyBulkCBuffer(c, node->name, CLUSTER_NAMELEN);
 
+    /* Add the additional endpoint information, this is all the known networking information
+     * that is not the preferred endpoint. */
+    void *deflen = addReplyDeferredLen(c);
+    int length = 0;
+    if (server.cluster_preferred_endpoint_type != CLUSTER_ENDPOINT_TYPE_IP) {
+        addReplyBulkCString(c, "ip");
+        addReplyBulkCString(c, node->ip);
```

Comment:
The type is "ip" for ipv4 and ipv6? I think it can be useful to fistinguidh then as "ipv4" etc.

Reply:
We don't really differentiate the different IP types today, I don't want to optimize earlier for that case. We can always extend the map to be ipv4 and ipv6 later.

Reply:
We can add "ipv4" and "ipv6" later but then we'll need to keep "ip" too for backward compat...?

Reply:
I think having that is "okay". We already show just an "ip" in multiple places and don't try to distinguish it between versions. We just give them an IP address they can talk to.

Reply:
Okay then, but we could also skip tagging them altogether. It's pretty easy for a client to tell a host name and an IP address apart.

Reply:
I don't want to push clients to know how to "parse" stuff, it's much easier to just do extra_networking['hostname'] as opposed to parsing all of them looking for a hostname.

Reply:
Fine, I got that. (Though they may need to parse the primary endpoint which isn't tagged...) The only point I wanted to make is that clients may still need to auto-detect if an IP is IPv4 or IPv6 and now we had a change to tag that information too, but if we don't differentiate between them in other places, then it's fine here too. I don't really prefer anything, just that the available options have been considered, and now they have.

I think the map for additional stuff is a very clean and nice format. It's probably future-proof too.

Reply:
That is a fair point. Statically typed languages will have to deal with it. Most dynamic languages usually just allow passing in either and it will figure it out. Most of the reason we're doing the new map instead of new fixed position fields is because many clients will pickup the hostname without any changes that way.

Reply:
Exactly, in C for example, you use `inet_pton()` with the `AF_INET` or the `AF_INET6` flag. Here's from Redis itself:

```C
int clusterStartHandshake(char *ip, int port, int cport) {
    clusterNode *n;
    char norm_ip[NET_IP_STR_LEN];
    struct sockaddr_storage sa;

    /* IP sanity check */
    if (inet_pton(AF_INET,ip,
            &(((struct sockaddr_in *)&sa)->sin_addr)))
    {
        sa.ss_family = AF_INET;
    } else if (inet_pton(AF_INET6,ip,
            &(((struct sockaddr_in6 *)&sa)->sin6_addr)))
    {
        sa.ss_family = AF_INET6;
    } else {
        errno = EINVAL;
        return 0;
    }
```